lib/unparser/preprocessor.rb

```diff
@@ -151,7 +151,53 @@ def chunked_children
 
     end # CollapseStrChildren
 
-    # Preprocessor eliminating unneded dstr nodes
+    # Preprocessor flattening unneeded dstr nesting. This happens when a dstr is generated from
+    # implicit string concatenation, and one of the strings being concatenated is itself a dstr.
+    #
+    # Example: "foo" "bar #{baz}"
+    #
+    # Without this preprocessor, this turns into: "foo#{"bar#{baz}"}"
+    # With this preprocessor, this turns into: "foobar #{baz}"
+    class FlattenImplicitDSTR < self
+
+      # No need to register :dsym, because Ruby doesn't do implicit concatenation of symbols
+      register :dstr
+
+      FLATTEN_CHILDREN = IceNine.deep_freeze([:dstr, :str])
+
+      # Return preprocessor result
+      #
+      # @return [Parser::AST::Node]
+      #
+      # @api private
+      #
+      def result
+        return node unless implicit_dstr?
+
+        flat_children = children.flat_map do |child|
+          child.type.equal?(:dstr) ? child.children : child
+        end
+        node.updated(nil, flat_children)
+      end
+
+    private
+
+      # Test for implicit dstr
+      #
+      # This should only ever be true for dstr nodes that are a result of implicit concatenation
+      # (see the comments on this class). Any other dstr node would have a :begin node as a child.
+      #
+      # @return [Boolean]
+      #
+      # @api private
+      #
+      def implicit_dstr?
+        children.map(&:type).all?(&FLATTEN_CHILDREN.method(:include?))
```

Comment:
Perhaps this isn't an issue, but this does create a temporary array, iterates over it, and then throws it away. It will probably be less memory usage if you did something like:

```ruby
children.all? { |child| FLATTEN_CHILDREN.include?(child.type) }
```

This also has the nice advantage of short-circuiting if one of the blocks return false, saving the need to fully iterate over all children.